src/github/notifications.ts

```diff
@@ -0,0 +1,381 @@
+			}
+			pullRequestModel.hasNotifications = false;
+			this._notifications.delete(identifier);
+			this._gitHubPrsTree.updateNotificationBadge(this);
+			this._gitHubPrsTree.refresh();
+		}
+
+	}
+
+	public async pollForNewNotifications() {
+		const response = await this.getNotifications();
+		if (response === undefined) {
+			return;
+		}
+		const { data, headers } = response;
+		const pollTimeSuggested = Number(headers['x-poll-interval']);
+
+		// Adapt polling interval if it has changed
+		if (pollTimeSuggested !== this._pollingDuration) {
+			this._pollingDuration = pollTimeSuggested;
+			if (this._pollingHandler && vscode.workspace.getConfiguration(SETTINGS_NAMESPACE).get(NOTIFICATION_SETTING)) {
+				Logger.appendLine('Notifications: Clearing interval');
+				clearInterval(this._pollingHandler);
+				Logger.appendLine(`Notifications: Starting new polling interval with ${this._pollingDuration}`);
+				this.startPolling();
+			}
+		}
+
+		// Only update if the user has new notifications
+		if (this._lastModified === headers['last-modified']) {
+			return;
+		}
+		this._lastModified = headers['last-modified'];
+
+		for (const prNotifications of this._notifications.values()) {
+			for (const notification of prNotifications) {
+				if (notification.pullRequestModel) {
+					notification.pullRequestModel.hasNotifications = false;
+				}
+			}
+		}
+		this._notifications.clear();
+
+		const currentRepos = new Map<string, GitHubRepository>();
+
+		this._reposManager.folderManagers.forEach(manager => {
+			manager.gitHubRepositories.forEach(repo => {
+				currentRepos.set(repo.remote.url, repo);
+			});
+		});
+
+		await Promise.all(data.map(async (notification) => {
+
+			const repoUrl = `${notification.repository.html_url}.git`;
+			const githubRepo = currentRepos.get(repoUrl);
+
+			if (githubRepo && notification.subject.type === 'PullRequest') {
+				const splitPrUrl = notification.subject.url.split('/');
+				const prNumber = splitPrUrl[splitPrUrl.length - 1];
+				const identifier = this.getPrIdentifier(notification);
+
+				const { remote, query, schema } = await githubRepo.ensure();
+
+				const { data } = await query<PullRequestState>({
+					query: schema.PullRequestState,
+					variables: {
+						owner: remote.owner,
+						name: remote.repositoryName,
+						number: Number(prNumber),
+					},
+				});
+
+				if (data.repository.pullRequest.state === 'OPEN') {
+
+					const newNotification = new Notification(
+						identifier,
+						Number(notification.id),
+						notification.reason,
+						notification.repository.name,
+						Number(prNumber)
+					);
+
+					const currentPrNotifications = this._notifications.get(identifier);
+					if (currentPrNotifications === undefined) {
+						this._notifications.set(identifier, [newNotification]);
+					}
+					else {
+						currentPrNotifications.push(newNotification);
+					}
+				}
+
+			}
+		}));
+
+		this.adaptPRNotifications();
+
+		this._gitHubPrsTree.updateNotificationBadge(this);
```

Comment:
This `updateNotificationBadge` call is circular as it immediately calls back into the `NotificaitonProvider`. Instead, can you use https://github.com/microsoft/vscode-pull-request-github/blob/ed64fa806143ba1a9cef6c593a1eda9feb24e7c3/src/view/prsTreeDataProvider.ts#L32-L33 to get the view and update the badge directly in `NotificationProvider`